authority/provisioner/scep.go

```diff
@@ -162,21 +179,97 @@ func (s *SCEP) Init(config Config) (err error) {
 	if s.MinimumPublicKeyLength == 0 {
 		s.MinimumPublicKeyLength = 2048
 	}
-
 	if s.MinimumPublicKeyLength%8 != 0 {
 		return errors.Errorf("%d bits is not exactly divisible by 8", s.MinimumPublicKeyLength)
 	}
 
+	// Set the encryption algorithm to use
 	s.encryptionAlgorithm = s.EncryptionAlgorithmIdentifier // TODO(hs): we might want to upgrade the default security to AES-CBC?
 	if s.encryptionAlgorithm < 0 || s.encryptionAlgorithm > 4 {
 		return errors.New("only encryption algorithm identifiers from 0 to 4 are valid")
 	}
 
+	// Prepare the SCEP challenge validator
 	s.challengeValidationController = newChallengeValidationController(
 		config.WebhookClient,
 		s.GetOptions().GetWebhooks(),
 	)
 
+	if decryptionKey := s.DecrypterKey; decryptionKey != "" {
```

Comment:
I started to look into this, and when a DB backs step-ca, we should store the key in the DB. We can also consider only supporting this case, even if the ca.json is used:

```json
{
  "...": "...",
  "decrypterKey": "LS0tLS1CRUdJTiBSU0EgUFJJVkFURSBLRVktLS0tLQo="
}
```

We must redact this and the password in the HTTP responses for provisioner and admin endpoints.

If we want to use files, we can consider an strategy used in the templates:

```go
type X509Options struct {
  Template string `json:"template,omitempty"`
  TemplateFile string `json:"templateFile,omitempty"`
  // ...
}
```

Reply:
Redaction can happen in the `api` package. See https://github.com/smallstep/certificates/pull/1414/files#diff-bf98d5fab5bcfabded9769069e7318683f549ec91cff84c5f87a9ae3a3ca3d8dR237. It's not super nice to have to replicate all the methods, but it doesn't rely on changing the values in the collection directly and also prevents refactoring the configuration of the provisioner vs. it being used.

Storing just the private key as bytes sounds OK to me, but we may need to support a password too? And there's also the case when it's provided as URI?

Having an option with and without a file sounds OK too.

Reply:
Perhaps something like this will support most of the cases:
```proto
message SCEPDecrypter {
    bytes decrypterCertificate = 1; // base64 cert
    bytes decrypterKey = 2;         // base64 key
    bytes decrypterKeyPassword = 3; // base64 (optional) password
    string decrypterKeyUri = 4;     // kmsuri
}
```

A URI can also be a file using `softkms:path/to/file.crt`. We can also add support to detect full/relative paths using `[/]path/to/file.crt`, or we can simplify it by adding a new `string decrypterKeyFile = 5;`

We might need a `KMS kms = 1;` like you have in the linkedca PR, but if we can avoid it, it would be great.

Reply:
The `linkedca.KMS` probably can be removed. It's what I tried first, but then the changes to parsing the KMS from an URI were introduced, and that's in use now to determine the type of KMS to use when configuring a key. There's a couple of special cases for `softkms` that I think can be improved, but I don't think they're blockers.

Having just the base64 cert is probably OK. That just works, and keeps thing simple in terms of storage, both in standalone (with and without remote administration enabled) and in hosted.

For the key, we would thus detect if the URI is set vs. the key is set, enforce and use only one of them? The key password could be empty, if the key URI has the appropriate fields set?